### DISCUSSION TITLE
GoRouter - Navigate using Riverpod

### DISCUSSION THREAD

**Discussion by rddewan** [rrousselGit/riverpod#1744] [Ideas]
I am have provider for GoRouter 
```
final goRouterProvider = Provider<GoRouter>((ref) { ...}
```

and now i can use anywhere inside my app as long as i have access to `ref` object
```
ref.read(goRouterProvider).push('/home');
```
I have a `class Notification` and it has a method to listen to a push notification 
```
void _handleMessage(RemoteMessage message) {
    switch (message.data['type']) {
      case 'product':        
        final String link = message.data['link'];     
        _ref.read(goRouterProvider).push(link); 
        break;
      default:
    }
    
  }
``` 
I am bit confuse here should I be pushing a route from this class ? since it's out of the widget , is it an anti-pattern or bad practice to navigate out of widget class?

Thank You

**Comment by bizz84** [comment]
I guess it's normal to do it like that when you're handling push notifications (since those are incoming events that are not triggered from the widgets).

I'm not sure if it's best practice though.

Also ask yourself if you should be using `push` or `go`. If the link represents a valid route match, maybe it makes sense to discard the navigation stack and just "go" to the new route.

**Reply by rddewan** [reply]
thank you for your answer
